File: examples/NavierStokes/SupersonicTGV/pdeapp_text2code.m
% Put the Exasim MATLAB frontend on the path. For an installed Exasim use
% run('<prefix>/share/exasim/matlab/exasim_setup.m') instead.
run(fullfile(fileparts(mfilename('fullpath')), '..', '..', '..', 'frontends', 'Matlab', 'exasim_setup.m'));

% initialize pde structure and mesh structure
[pde,mesh] = initializeexasim();

% Define a PDE model: governing equations, initial solutions, and boundary conditions
pde.model = "ModelD";          % ModelC, ModelD, ModelW
pde.modelfile = "pdemodel";    % name of a file defining the PDE model

% Choose computing platform and set number of processors
%pde.platform = "gpu";         % choose this option if NVIDIA GPUs are available
pde.mpiprocs = 12;             % number of MPI processors
pde.hybrid = 1;

% Set discretization parameters, physical parameters, and solver parameters
pde.porder = 2;          % polynomial degree
pde.pgauss = 2*pde.porder;
pde.torder = 3;          % time-stepping order of accuracy
pde.nstage = 3;          % time-stepping number of stages

% Lusher and Sandham, AIAA J. 2020: Mref = 1.25 TGV, fixed
% nondimensional time step Delta t = 5e-4, advanced to t = 20.
Deltat = 1e-2;
nsteps = round(20/Deltat);
pde.dt = Deltat*ones(1,nsteps);   % time step sizes
pde.saveSolFreq = 20;
pde.saveSolOpt = 0;
% [8e-3, 4e-3, 2e-3, 1.6e-3, 1.5e-3, 1.4e-3, 1.34e-3]

nspatial = 16;
gam = 1.4;                      % specific heat ratio
Re = 1600;                      % Reynolds number
Pr = 0.71;                      % Prandtl number    
Minf = 1.25;                    % reference Mach number
rhoRef = 1.0;                   % nondimensional reference density
hm = 2*pi/nspatial;             % AV sensor length scale
avcoeff = 2.0e-3;               % AV coefficient
pde.physicsparam = [gam Re Pr Minf rhoRef hm avcoeff pde.porder];
pde.tau = 5.0;                  % DG stabilization parameter
pde.GMRESortho = 1;
pde.GMRESrestart=24;
pde.linearsolvertol=1e-7;
pde.linearsolveriter=24;
pde.preconditioner=1;
pde.precMatrixType=2;
pde.NLiter=1;
pde.NLtol = 1e-8;
pde.ppdegree = 0;
pde.RBdim = 5;
pde.gencode = 1;

% Artificial viscosity for shock capturing; pdemodel uses av = v(1).
pde.AV = 1;
pde.frozenAVflag = 1;
pde.AVsmoothingIter = 2;

% Create a periodic cube 0 <= x,y,z <= 2*pi*L with L = 1.
[mesh.p,mesh.t] = cubemesh(nspatial,nspatial,nspatial,1);
mesh.p = 2*pi*mesh.p;
% expressions for domain boundaries
mesh.boundaryexpr = {@(p) abs(p(2,:))<1e-8, @(p) abs(p(1,:)-2*pi)<1e-8, @(p) abs(p(2,:)-2*pi)<1e-8, @(p) abs(p(1,:))<1e-8, @(p) abs(p(3,:))<1e-8, @(p) abs(p(3,:)-2*pi)<1e-8};
mesh.boundarycondition = [1;1;1;1;1;1];
% Set periodic boundary conditions
mesh.periodicexpr = {2, @(p) p([2 3],:), 4, @(p) p([2 3],:); 1, @(p) p([1 3],:), 3, @(p) p([1 3],:); 5, @(p) p([1 2],:), 6, @(p) p([1 2],:)};

exasimroot = fullfile(fileparts(mfilename('fullpath')), '..', '..', '..');
exportdir = fullfile(exasimroot, 'apps', 'navierstokes', 'supersonicTGV');
exporttext2code(pde, mesh, exportdir);

%[sol,pde,mesh] = exasim(pde,mesh);

File: examples/NavierStokes/SupersonicTGV/pdeapp_text2code.py
"""Export the Supersonic Taylor-Green vortex as a Text2Code app.

This is the Python equivalent of ``pdeapp_text2code.m``. It writes the
Text2Code package to ``apps/navierstokes/supersonicTGV`` without running the
solver.
"""

from pathlib import Path
import sys

import numpy

exasimroot = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(exasimroot / "frontends" / "Python"))

import exasim


def main():
    pde, mesh = exasim.initializeexasim()

    pde["model"] = "ModelD"
    pde["modelfile"] = "pdemodel"
    pde["ncu"] = 5
    pde["nco"] = 1

    pde["mpiprocs"] = 12
    pde["hybrid"] = 1

    pde["porder"] = 2
    pde["pgauss"] = 2 * pde["porder"]
    pde["torder"] = 3
    pde["nstage"] = 3

    deltat = 1.0e-2
    nsteps = round(20.0 / deltat)
    pde["dt"] = deltat * numpy.ones(nsteps)
    pde["saveSolFreq"] = 20
    pde["saveSolOpt"] = 0

    nspatial = 16
    gam = 1.4
    Re = 1600.0
    Pr = 0.71
    Minf = 1.25
    rhoRef = 1.0
    hm = 2.0 * numpy.pi / nspatial
    avcoeff = 2.0e-3
    pde["physicsparam"] = numpy.array([gam, Re, Pr, Minf, rhoRef, hm, avcoeff, pde["porder"]])

    pde["tau"] = numpy.array([5.0])
    pde["GMRESortho"] = 1
    pde["GMRESrestart"] = 24
    pde["linearsolvertol"] = 1.0e-7
    pde["linearsolveriter"] = 24
    pde["preconditioner"] = 1
    pde["precMatrixType"] = 2
    pde["NLiter"] = 1
    pde["NLtol"] = 1.0e-8
    pde["ppdegree"] = 0
    pde["RBdim"] = 5
    pde["gencode"] = 1

    pde["AV"] = 1
    pde["frozenAVflag"] = 1
    pde["AVsmoothingIter"] = 2

    mesh["p"], mesh["t"] = exasim.Mesh.cubemesh(nspatial, nspatial, nspatial, 1)[0:2]
    mesh["p"] = 2.0 * numpy.pi * mesh["p"]
    mesh["t"] = mesh["t"] + 1

    mesh["boundaryexpr"] = [
        "abs(y)<1e-8",
        "abs(x-2*pi)<1e-8",
        "abs(y-2*pi)<1e-8",
        "abs(x)<1e-8",
        "abs(z)<1e-8",
        "abs(z-2*pi)<1e-8",
    ]
    mesh["boundarycondition"] = numpy.array([1, 1, 1, 1, 1, 1])

    # Python exporttext2code consumes mesh["periodicboundary"] and expands
    # xy/xz/yz into the text2code coordinate lists used by the MATLAB exporter.
    mesh["periodicboundary"] = [
        [2, "yz", 4, "yz"],
        [1, "xz", 3, "xz"],
        [5, "xy", 6, "xy"],
    ]

    exportdir = exasimroot / "apps" / "navierstokes" / "supersonicTGV"
    exasim.exporttext2code(pde, mesh, exportdir)


if __name__ == "__main__":
    main()
